### PR TITLE
feat: implement namespaced state model to support local and global states

### LINKED ISSUE #71: Implement namespaced State
### Requested Feature

Currently all steps in a workflow share the same state. It causes problem where steps can overwrite each others state unintentionally (e.g. using the same key). We need to have the concept of "namespaced" state such as "local" or "global" so that when a state is setting or getting value, it should be able to set and retrieve correctly.

step.State().WithNamespace(name).Set("KEY1", "Value0") // 
step.State().Global().Set("KEY1", "Value1") // 
step.State().Local().Set("KEY1", "Value2") // set "KEY1" in local state without polluting global
step.State().Global.Get("KEY1) // return "Value1" instead of "Value2"

### Example
```
// Each step uses its own namespace (isolated)
func setupBindMount(name string, target string) automa.Builder {
    return automa.NewStepBuilder().
        WithExecute(func(ctx context.Context, stp automa.Step) *automa.Report {
            bindMount := mount.BindMount{
                Source: path.Join(core.Paths().SandboxDir, target),
                Target: target,
            }
            
            // Option 1: use custom namespace per step
            stp.State().WithNamespace(name).Set("bind-mount", bindMount)
            
            // Option 2: use local namespace (recommended)
            stp.State().Local().Set("bind-mount", bindMount)
            
            // Read shared config from global
            env := stp.State().Global().String("env")
            
            // ... rest of logic
        }).
        WithRollback(func(ctx context.Context, stp automa.Step) *automa.Report {
            // Read from same namespace
            var bindMount mount.BindMount
            if val, ok := stp.State().Local().Get("bind-mount"); ok {
                bindMount = val.(mount.BindMount)
            }
            // ... rollback logic
        })
}
```

### Business Need

A step implementation doesn't need to worry about state overwrite. It needs clear semantics of setting and getting value from state.

### Additional Context
Workflow needs to pass namespaced step into each step before execution where the global state would be shared accross all steps in a workflow.

### PR DESCRIPTION
**Pull request by leninmehedy**
## Description

This pull request introduces namespace-aware state management for workflow steps, enhancing both flexibility and thread-safety in state handling. The changes replace the previous state bag interface with a new `NamespacedStateBag`, allowing separation of local, global, and custom state namespaces. Additionally, the implementation and tests ensure robust concurrent access and merging, and documentation has been added to clarify state preservation configuration.

Namespace-aware state management:

* Introduced the `NamespacedStateBag` interface, supporting local, global, and custom namespaces for step state isolation and sharing. The `Step` interface and related builder methods now use `NamespacedStateBag` instead of `StateBag`. (`automa.go`, `step_builder.go`) [[1]](diffhunk://#diff-2f78a604a2a388d0cd14602cb33e14d37d22b5a98b276f6774a39cbcd2c49d83L13-R21) [[2]](diffhunk://#diff-de7fbc5fe7dcf3fe1020642b6d29bc5047a74e96a7db21e1157221bc840cc214L65-R65)
* Added `SyncNamespacedStateBag` as the thread-safe implementation, including methods for lazy initialization, deep cloning, and merging of namespaces. (`state_namespaced.go`)

Thread-safety and concurrency improvements:

* Comprehensive concurrency tests for `SyncNamespacedStateBag` covering local, global, custom, clone, and merge operations to ensure thread-safe state management. (`state_namespaced_concurrency_test.go`)

State bag interface and method updates:

* Updated `StateBag.Merge` and `StateBag.Clone` to return errors, improving error handling for state operations. (`automa.go`, `state.go`, `state_test.go`) [[1]](diffhunk://#diff-2f78a604a2a388d0cd14602cb33e14d37d22b5a98b276f6774a39cbcd2c49d83L25-R32) [[2]](diffhunk://#diff-320959764bfa2277e0bdd4eaa741f02891d123c42150d4ccb76de0bdbf6e591aL78-R87) [[3]](diffhunk://#diff-864944c880a2e4db1b8f93f33a58303d840fb4b6c50147cc2f915c6d48f7bf2cL127-R128) [[4]](diffhunk://#diff-864944c880a2e4db1b8f93f33a58303d840fb4b6c50147cc2f915c6d48f7bf2cL136-R138)
* Step builder now clones the entire `NamespacedStateBag` (all namespaces) when copying steps, ensuring isolation between workflow instances. (`step_builder.go`)

Documentation:

* Added a new guide explaining state preservation configuration, including memory and rollback tradeoffs for workflows. (`docs/state-preservation.md`)

### `StateBag.Merge()` signature changed
**Before:** `Merge(other StateBag) StateBag`  
**After:** `Merge(other StateBag) (StateBag, error)`

### `State()` / `WithState()` now use `NamespacedStateBag`
**Before:** `State() StateBag`  
**After:** `State() NamespacedStateBag`

### Migration
Replace direct state access:
```go
// Before
step.State().Set("key", value)
step.State().Get("key")

// After (local namespace)
step.State().Local().Set("key", value)
step.State().Local().Get("key")

// After (shared across steps)
step.State().Global().Set("key", value)
```

### Related Issues

* Closes #71 
* Related to #

### Pull request (PR) checklist

* \[ ] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[ ] This PR added no TODOs or commented out code
* \[ ] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[ ] Any `go.mod` and/or `go.sum` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[ ] All changes in this PR are included in the description
* \[ ] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit
  message guidelines' below have been followed

### Testing

* \[ ] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* TBD

The following was not tested:

* TBD

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a
   breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any
   type. NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: workflow.go [hunk 1]
```diff
@@ -221,39 +276,53 @@ func (w *workflow) Execute(ctx context.Context) *Report {
...
 
-		// attach snapshot for possible rollback (keeps alignment even if nil)
-		stepStates = append(stepStates, stepState)
-
 		// make sure step has its state before calling Prepare so Prepare can access it.
 		// It also ensures during Execute the step has the correct state
-		step = step.WithState(stepState)
+		if stepState != nil {
+			step = step.WithState(stepState)
+		}
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go]: changed to a map

**Review by brunodam — APPROVED** [review]
Thanks, Lenin. LGTM

File: workflow.go [hunk 2]
```diff
@@ -285,6 +363,30 @@ func (w *workflow) Execute(ctx context.Context) *Report {
...
+						Str("stepId", step.Id()).
+						Msg("failed to clone state for rollback snapshot; using current state reference (may be mutated by later steps before rollback)")
+
+					// Store non-cloned state for rollback
+					stepStates[step.Id()] = state
+				} else {
+					stepStates[step.Id()] = clonedState
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go]: Why this is correct:
- preserveStatesForRollback = false means "I don't need state for rollback"

Use cases:
- Rollback steps are idempotent (don't need state)
- No rollback needed at all
- Memory optimization when rollback doesn't require state

If rollback does need state, user should set preserveStatesForRollback = true
